distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso

```diff
@@ -0,0 +1,188 @@
+            Value_Type.Char size variable ->
+                case variable of
+                    True  ->
+                        is_unbounded = case size of
+                            Nothing -> True
+                            Positive_Integer.Value integer -> integer == max_precision
+                        case is_unbounded of
+                            True  -> SQL_Type.Value Types.VARCHAR "NVARCHAR(MAX)"
+                            False -> SQL_Type.Value Types.VARCHAR "NVARCHAR" size
+                    False -> SQL_Type.Value Types.CHAR "NCHAR" size
+            Value_Type.Time -> SQL_Type.Value Types.TIME "TIME"
+            Value_Type.Date -> SQL_Type.Value Types.DATE "DATE"
+            Value_Type.Date_Time with_timezone ->
+                type_name = if with_timezone then "DATETIMEOFFSET" else "DATETIME2"
+                SQL_Type.Value Types.TIMESTAMP type_name
+            Value_Type.Binary size variable_length ->
+                ## Max size for BINARY and VARBINARY is 8000, so we have to use VARBINARY(MAX) for larger sizes.
+                need_varbinary_max = size.is_nothing || (size > 8000)
+                case need_varbinary_max of
+                    True  -> SQL_Type.Value Types.BINARY "VARBINARY(MAX)"
+                    False -> case variable_length of
+                        True  -> SQL_Type.Value Types.BINARY "VARBINARY" size
+                        False -> SQL_Type.Value Types.BINARY "BINARY" size
+            Value_Type.Mixed -> Error.throw (Unsupported_Database_Operation.Error "SQLServer tables do not support Mixed types.")
+            Value_Type.Unsupported_Data_Type type_name underlying_type ->
+                underlying_type.if_nothing <| Error.throw <| Illegal_Argument.Error <|
+                    "An unsupported SQL type ["+type_name.to_text+"] cannot be converted into an SQL type because it did not contain the SQL metadata needed to reconstruct it."
+
+        approximated_value_type = SQLServer_Type_Mapping.sql_type_to_value_type result
+        problems = if approximated_value_type == value_type then [] else [Inexact_Type_Coercion.Warning value_type approximated_value_type]
+        on_problems.attach_problems_before problems result
+
+    ## PRIVATE
+    sql_type_to_value_type : SQL_Type -> Value_Type
+    sql_type_to_value_type sql_type =
+        simple_type = case sql_type.typeid of
+            Types.BIT -> Value_Type.Boolean
+            Types.TINYINT -> Value_Type.Integer Bits.Bits_16
+            Types.SMALLINT -> Value_Type.Integer Bits.Bits_16
+            Types.INTEGER -> Value_Type.Integer Bits.Bits_32
+            Types.BIGINT -> Value_Type.Integer Bits.Bits_64
+            ## SQLServer real is 32-bit, but our In Memory engine only has 64-bit floating point numbers so we have to upcast it.
+            Types.REAL -> Value_Type.Float Bits.Bits_64
+            Types.DOUBLE -> Value_Type.Float Bits.Bits_64
```

Comment:
This is not the right approach.

A dialect's type mapping deals with how the SQL types map to our `Value_Type` representation of the same concept. They should reflect what is really there in the Database. So if the Database supports 32-bit floats, they should map to `Value_Type.Float ..Bits_32` - that is exactly the purpose of that type.

In the current state - if I load in a SQL Server table that has a `REAL` column - it will wrongly report its type as 64-bit float - but it really is 32-bit.

This is the mapping of types itself. Now, how these types are handled during materialization to memory is a different story - I think our `double_fetcher` will automatically materialize either variant of `Float` as 64-bit floats that are the only ones we currently support in-memory.

Reply:
Fixed